### PR TITLE
Try to fix #35

### LINKED ISSUE #35: Buffer overflow in libfuse patch
> 	// For i = 4...99, check if there is a binary called "fusermount" + i
> 	// It is not yet known whether this will work for our purposes, but it is better than not even attempting
> 	for (int i = 4; i < 100; i++) {
> 		prog = findBinaryInFusermountDir("fusermount" + i);
> 		if (access(prog, X_OK) == 0)
> 			return prog;
> 	}

This adds to the string's pointer, it does not make a string concatenation at all. e.g it will lead to the following tries:
 - "rmountXXXX"
 - "mountXXXXX"
 - etc...
 With Xs being invalid memory addresses.

### PR DESCRIPTION
**Pull request by probonopd**
Closes #35
Thanks @Bqleine

### CODE REVIEW

**Comment by probonopd** [comment]
Why...

```
+ patch -p1
patching file lib/mount.c
patch: **** malformed patch at line 96:  		_exit(1);
```

**Comment by probonopd** [comment]
Build [artifacts](https://github.com/AppImage/type2-runtime/actions/runs/9767294099/artifacts/1661245349) for testing.

File: patches/libfuse/mount.c.diff [hunk 1]
```diff
@@ -65,8 +63,10 @@ index d71e6fc55..acc1711ff 100644
 +	// For i = 4...99, check if there is a binary called "fusermount" + i
 +	// It is not yet known whether this will work for our purposes, but it is better than not even attempting
 +	for (int i = 4; i < 100; i++) {
-+		prog = findBinaryInFusermountDir("fusermount" + i);
-+		if (access(prog, X_OK) == 0)
++		char binaryName[20];
```
Comment (Bqleine) [inline patches/libfuse/mount.c.diff:66]: Why twenty? It's at most thirteen characters including the null character, or twenty-one if you want to fit every integer.

Comment (probonopd) [reply patches/libfuse/mount.c.diff]: Agree. 13 it is.

**Comment by probonopd** [comment]
Actually... won't

https://github.com/AppImage/type2-runtime/pull/32/files

fix this once it finally gets reviewed and merged?

File: patches/libfuse/mount.c.diff [hunk 2]
```diff
@@ -65,8 +63,10 @@ index d71e6fc55..acc1711ff 100644
 +	// For i = 4...99, check if there is a binary called "fusermount" + i
 +	// It is not yet known whether this will work for our purposes, but it is better than not even attempting
 +	for (int i = 4; i < 100; i++) {
-+		prog = findBinaryInFusermountDir("fusermount" + i);
-+		if (access(prog, X_OK) == 0)
++		char binaryName[13];
++		sprintf(binaryName, "fusermount%d", i);
++		prog = findBinaryInFusermountDir(binaryName);
++		if (prog && access(prog, X_OK) == 0)
```
Comment (TheAssassin) [inline patches/libfuse/mount.c.diff:69]:
```suggestion
+		if (prog != NULL && access(prog, X_OK) == 0)
```

**Review by TheAssassin — CHANGES_REQUESTED** [review]
Had a first glance. Please revisit, I'll have another look afterwards.

File: patches/libfuse/mount.c.diff [hunk 3]
```diff
@@ -65,8 +63,10 @@ index d71e6fc55..acc1711ff 100644
 +	// For i = 4...99, check if there is a binary called "fusermount" + i
 +	// It is not yet known whether this will work for our purposes, but it is better than not even attempting
 +	for (int i = 4; i < 100; i++) {
-+		prog = findBinaryInFusermountDir("fusermount" + i);
-+		if (access(prog, X_OK) == 0)
++		char binaryName[13];
++		sprintf(binaryName, "fusermount%d", i);
```
[the remark this answers is not in the document]
Comment (probonopd) [reply patches/libfuse/mount.c.diff:67]: `fusermount `... `fusermount99 `is what I had in mind

[the remark this answers is not in the document]
Comment (probonopd) [reply patches/libfuse/mount.c.diff]: Not sure what you mean with the further check - in which case would this even be needed?

[the remark this answers is not in the document]
Comment (TheAssassin) [reply patches/libfuse/mount.c.diff]: If `i` overflows for whatever reason (might be future changes), `snprintf` will not render additional digits but it'd be unexpected to run into the limit anyway. It's good practice to check the return value. See, e.g., https://stackoverflow.com/questions/54084570/how-to-detect-snprintf-errors.

**Comment by Bqleine** [comment]
The more I look into this, the more it seems like a bad idea.

Firstly, @TheAssassin noticed the security issues in your patch and told you, but you merged the patch without more discussion. https://github.com/AppImage/type2-runtime/pull/11#issuecomment-1616169180

Then in the patch you wrote that « For security reasons, we do not search the binary on the $PATH; », but now we are doing it in #32?

I don't understand everything but we should seriously reconsider the security implications of doing such a patch. And ask libfuse's developers' opinion, as they either have a good reason of not doing it or could merge it directly upstream and give it a proper review for security issues.

Because putting buffer overflows in someone else's code is not ok.

**Comment by TheAssassin** [comment]
I wish I had more time to work on AppImage and all of my other projects because I know I could handle most of this better, however that's likely not going to change soon. This is a tiny project team wise but has a large impact on users and developers, which is perfectly illustrated by https://xkcd.com/2347/ (change Nebraska to Germany).

**Pull request closed without merging**